package.xml

```diff
@@ -38,12 +38,13 @@
 
   <build_depend>git</build_depend>
   <build_depend>google-mock</build_depend>
-  <build_depend>python-sphinx</build_depend>
+  <build_depend>gtest</build_depend>
+  <build_depend>python3-sphinx</build_depend>
 
-  <depend>boost</depend>
-  <depend>ceres-solver</depend>
+  <depend>libboost-iostreams-dev</depend>
   <depend>eigen</depend>
   <depend>libcairo2-dev</depend>
+  <depend>libceres-dev</depend>
```

Comment:
This does not work. For example Kinetic is still supported until April 2021 which does not work with this change. How about making this change only for ROS2 and keeping the `ceres-solver` for ROS1?

Reply:
This isn't actually a ROS 1 vs. ROS 2 thing.  It's more of a Linux distribution thing.

For what it is worth, the `libceres-dev` rosdep key is available for Kinetic (Ubuntu distribution Xenial): https://github.com/ros/rosdistro/blob/d8a1a79b5958aaac468d665cb99eb4c3d68c4b09/rosdep/base.yaml#L1950

If we look in Ubuntu, we see that that maps to this package version: https://packages.ubuntu.com/xenial/libceres-dev .

What I don't know off-hand is whether libceres-dev 1.11.0 is new enough to compile the latest cartographer against.  Do you happen to know?

Reply:
The reason was that `libceres-dev` in Xenial was packaged incorrectly and does not contain the CMake config files. Later Ubuntu versions have this, so we use `libceres-dev` starting at Bionic.

Reply:
All right, that makes sense.

The bit that still doesn't quite make sense to me is the `ceres-solver` key, though.  How is that key being satisfied for Xenial?  There is no entry in https://github.com/ros/rosdistro/blob/master/rosdep/base.yaml or https://github.com/ros/rosdistro/blob/master/kinetic/distribution.yaml for it.

Reply:
Xenial is anyways EOL in 7 months (April 2021) - maybe we just drop support early for Xenial and for anybody who is interested in Xenial we make a [tag](https://github.com/cartographer-project/cartographer/tags) prior to merging this?

Reply:
Let's not do this:
1. Ubuntu Xenial and ROS Kinetic are still supported. I think the most important CI this needs to pass is the one in cartographer_ros for the various ROS1 versions still supported.
2. What is the rationale behind this change? We build Ceres for compatibility with old versions for now, but is this actually causing any issues in ROS2? I think the focus of this PR should strictly be adding support for ROS2 in this PR. Other changes should go to separate PRs.

Reply:
> * What is the rationale behind this change? We build Ceres for compatibility with old versions for now, but is this actually causing any issues in ROS2? I think the focus of this PR should strictly be adding support for ROS2 in this PR. Other changes should go to separate PRs.

This is causing problems in ROS, period.  As it is, we cannot release cartographer into ROS 1 or ROS 2.  The reason is that the ROS buildfarm requires that all dependency keys be resolvable to actual package names (either from the system packages, or from elsewhere in the ROS ecosystem).  Since `ceres-solver` doesn't resolve to anything in ROS 1 or ROS 2, we can't release it.

We get around this currently by patching the ceres-solver key out when doing releases for ROS 1 (see https://github.com/ros-gbp/cartographer-release/blob/master/kinetic/package.xml for the package.xml we "overlay" on the sources), and by having a forked cartographer for ROS 2 (https://github.com/ros2/cartographer, the fork we are currently trying to get rid of).  However, the best way to solve this is to make the key resolvable in the upstream, which is what this patch is trying to do.

>     1. Ubuntu Xenial and ROS Kinetic are still supported. I think the most important CI this needs to pass is the one in cartographer_ros for the various ROS1 versions still supported.

I looked into how this is working for Kinetic a bit more.  We are hosting an upgraded version of libceres on our apt repository: http://packages.ros.org/ros-testing/ubuntu/pool/main/c/ceres-solver/ .  With that in place, we are actually able to resolve the `libceres-dev` key to a proper package name on Xenial, and hence we can release cartographer into Kinetic.  Thus, I believe that CI will continue to work with Xenial even after this change (travis seems to agree with me here).

Reply:
Thanks for the context, this really helps everyone understand what we are looking into here. I verified that we can install the ROS-provided libceres-dev to make it work. Unfortunately when I start a docker build for kinetic, I get an error. Could it be that the current package.xml is missing a dependency? The error unfortunately does not mention which file is missing:
```
 OSError: Failed command '['/opt/cartographer_ros/env.sh', 'cmake', '/catkin_ws/src/cartographer_ros/cartographer_ros_msgs', '-DCATKIN_DEVEL_PREFIX=/catkin_ws/devel_isolated/cartographer_ros_msgs', '-DCMAKE_INSTALL_PREFIX=/opt/cartographer_ros', '-G', 'Ninja']': [Errno 2] No such file or directory
```

I think we need to investigate more. Once this is resolved, I think it is okay to break existing users of cartographer_ros with such a change as we then know how to fix it quickly in cartographer_ros. Without a follow up PR already tested and ready to go, I think we would let down existing users.

Different topic: I might be mistaken, but it seems to me that you are frustrated about the slow progress ("This is causing problems in ROS, period.") but given that everyone here is contributing voluntarily, this will only proceed when someone is willing to put in the effort to move it forward, as I just did the last hour in trying it out to see whether it works. I think this is the only blocker really and to my knowledge everyone would love to see ROS2 supported as soon as we can.

Reply:
My investigation so far mostly for Kinetic:
1. All the Catkin packages in our ROS CI use `/opt/cartographer_ros/env.sh`. Not sure where this is configured/determined and why. Does someone know?
2. If we build Ceres first which is a non-Catkin package, it install it, so everything works.
3. If we do not build Ceres we fail, but if we build Cartographer first it would fix it.
4. Not sure whether this would fix all issues, especially when building outside of CI. After all we do not want to fix CI by hiding existing issues. Any one understands enough of Catkin to know what is going on?
5. Melodic/Noetic: also stops working, same problem. What I do not understand and have not investigated yet: why do  they work before but not with the change? It seems before they somehow know not to use the non-existing `env.sh`. 

Otherwise, my suggestion would be the following: on Monday I create a PR that fixes Cartographer ROS by reordering the package builds and verify it works. Then we merge the Cartographer changes, quickly followed by the Cartographer ROS PR.

Reply:
> Thanks for the context, this really helps everyone understand what we are looking into here.

You are welcome, happy to provide it.

> Could it be that the current package.xml is missing a dependency? The error unfortunately does not mention which file is missing:

Hm, that seems like it is either missing cmake, or the initial `/opt/cartographer_ros/env.sh` file.  Not sure what is going on there.

> I think we need to investigate more. Once this is resolved, I think it is okay to break existing users of cartographer_ros with such a change as we then know how to fix it quickly in cartographer_ros. Without a follow up PR already tested and ready to go, I think we would let down existing users.

Sorry, I don't think I understand this bit.  I don't *think* that making the change in this PR will cause any problems for downstream users, but maybe I'm misunderstanding.

> Different topic: I might be mistaken, but it seems to me that you are frustrated about the slow progress ("This is causing problems in ROS, period.")

No, I'm not frustrated.  I was just emphasizing that we need this change in both ROS 1 and ROS 2.  No worries, I am doing this on volunteer time too :).

Reply:
Okay. I think I understand it now:
1. We build package explicitly in a specific order in our cartographer_ros CI to allow Docker to cache results. This has the side effect that if the order which Catkin chooses changes, we have to adapt CI.
2. When we remove the Ceres packages Catkin changes the order in which packages are built which breaks CI. This is even true for the newer ROS versions where the package is just present but was not actually compiled by CI.
3. Especially we have to use the ordering Catkin uses, not just any ordering which would work according to the dependencies. In our case, we must built cartographer before cartographer_ros_msgs even though the latter does not depend on the former.

Reply:
Resolved - we depend on libceres-dev now (hence no diff anymore)